File: usaco/section_3.1/inflate.py
"""
ID: huangch7
LANG: PYTHON3
TASK: inflate
"""


task_name = "inflate"

fin = open(f'{task_name}.in', 'r')
fout = open(f'{task_name}.out', 'w')

M, N = [int(x) for x in fin.readline().strip().split()]

info = []

for _ in range(N):
    point, time = [int(x) for x in fin.readline().strip().split()]
    info.append((point, time))

# print(info)
# dp[v] 表示容量为v能的到的最大分
dp = [None] * (M + 1)
dp[0] = 0

for i in range(1, M + 1):
    for j in range(len(info)):
        prev = i - info[j][1]
        if prev < 0:
            continue
        if dp[i] is None or dp[i] < dp[prev] + info[j][0]:
            dp[i] = dp[prev] + info[j][0]
    # 可以少装一些
    if dp[i] is None or dp[i] < dp[i - 1]:
        dp[i] = dp[i - 1]
# for i in range(0, M + 1, 10):
#     print(i, dp[i: i + 10])
fout.write(f"{dp[M]}\n")
fout.close()

File: usaco/training/section_3.1/inflate.cpp
/*
ID: huangch7
LANG: C++
TASK: inflate
*/

#include <iostream>
#include <vector>
#include <fstream>
#include <algorithm>

using namespace std;

int main() {
    ifstream fin("inflate.in");
    ofstream fout("inflate.out");

    int M, N;
    fin >> M >> N;

    vector<pair<int, int> > info(N);

    for (int i = 0; i < N; ++i) {
        int point, time;
        fin >> point >> time;
        info[i] = make_pair(point, time);
    }

    // dp[v] 表示容量为v能的到的最大分
    vector<int> dp(M + 1, 0);

    for (int i = 1; i <= M; ++i) {
        for (int j = 0; j < N; ++j) {
            int prev = i - info[j].second;
            if (prev >= 0) {
                dp[i] = max(dp[i], dp[prev] + info[j].first);
            }
        }
    }

    fout << dp[M] << endl;
    fout.close();

    return 0;
}
